cpp/src/fil/infer.cu

```diff
@@ -504,6 +488,41 @@ void infer_k_launcher(storage_type forest, predict_params params,
   CUDA_CHECK(cudaPeekAtLastError());
 }
 
+template <leaf_algo_t leaf_algo, typename storage_type>
+void infer_k_launcher(storage_type forest, predict_params params,
+                      cudaStream_t stream, int blockdim_x) {
+  // TODO(levsnv): move this to forest init()
+  /** searching for the most items per block while respecting the shared
+  * memory limits creates a full linear programming problem.
+  * solving it in a single equation looks less tractable than this */
+  params.cols_in_shmem = true;
+  int num_items = 0;
+  size_t shm_sz;
+  try_nitems<leaf_algo, 1>(&num_items, &shm_sz, params);
+  if (num_items == 0) {
+    params.cols_in_shmem = false;
+    try_nitems<leaf_algo, 1>(&num_items, &shm_sz, params);
+    try_nitems<leaf_algo, 2>(&num_items, &shm_sz, params);
+    try_nitems<leaf_algo, 3>(&num_items, &shm_sz, params);
+    try_nitems<leaf_algo, 4>(&num_items, &shm_sz, params);
+    ASSERT(num_items != 0, "FIL out of shared memory. >>5'000 classes?");
```

Comment:
Could are more specific error message be provided, e.g. that the maximum supported number of classes is exceeded?